programs/stake/src/stake_state.rs

```diff
@@ -761,11 +761,22 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
             match self.state()? {
                 StakeState::Stake(meta, mut stake) => {
                     meta.authorized.check(signers, StakeAuthorize::Staker)?;
+                    let original_lamports_per_byte_year = calculate_lamports_per_byte_year(
+                        self.data_len()? as u64,
+                        meta.rent_exempt_reserve,
+                    );
 
                     // verify enough lamports for rent in new stake with the split
                     if split.lamports()? + lamports < meta.rent_exempt_reserve
-                     // verify enough lamports left in previous stake and not full withdrawal
+                    // verify enough lamports left in previous stake and not full withdrawal
                         || (lamports + meta.rent_exempt_reserve > self.lamports()? && lamports != self.lamports()?)
+                    // verify that the same rent_exempt_reserve is sufficient to make the new
+                    // stake account rent exempt needed because the new uninitialized may be any
+                    // size larger than std::mem::size_of::<StakeState>()
+                        || calculate_rent_exempt_reserve(
+                            split.data_len()? as u64,
+                            original_lamports_per_byte_year,
+                        ) > meta.rent_exempt_reserve
                     {
                         return Err(InstructionError::InsufficientFunds);
                     }
```

Comment:
We need to use the split's `rent_exempt_reserve` below, no? (Sorry can't comment outside diff context)

Reply:
No, the rent_exempt_reserve needs to remain the same between the source and split accounts. We just don't want to allow that to put the split account in a non-exempt state

Reply:
Mm, maybe we can actually, as long as the stake being split never increases as a result.

Reply:
Updated. No tests failed, so that's good. I still need to add a case for larger source to smaller split, splitting partial funds.